### ISSUE TITLE
Support query parameters with special characters

### ISSUE THREAD

**Issue by rwinch** [thymeleaf/thymeleaf#23]
I have a use case where I need to be able to create a URL with [] in the parameter name (i.e. /context/url?a[0]=b). However, when I create a link like:

``` xml
<a href="#" th:href="@{/url(a[0]='b')}">List</a>
```

I get an Exception

``` java
org.thymeleaf.exceptions.TemplateProcessingException: Could not parse as expression: "@{/url(a[0]='b')}" (home:9)
    org.thymeleaf.standard.expression.StandardExpressionParser.parseExpression(StandardExpressionParser.java:128)
    org.thymeleaf.standard.expression.StandardExpressionParser.parseExpression(StandardExpressionParser.java:70)
    org.thymeleaf.standard.expression.StandardExpressionProcessor.parseExpression(StandardExpressionProcessor.java:53)
    org.thymeleaf.standard.expression.StandardExpressionProcessor.processExpression(StandardExpressionProcessor.java:91)
    org.thymeleaf.standard.processor.attr.AbstractStandardSingleAttributeModifierAttrProcessor.getTargetAttributeValue(AbstractStandardSingleAttributeModifierAttrProcessor.java:59)
    org.thymeleaf.processor.attr.AbstractSingleAttributeModifierAttrProcessor.getModifiedAttributeValues(AbstractSingleAttributeModifierAttrProcessor.java:59)
    org.thymeleaf.processor.attr.AbstractAttributeModifierAttrProcessor.processAttribute(AbstractAttributeModifierAttrProcessor.java:61)
    org.thymeleaf.processor.attr.AbstractAttrProcessor.doProcess(AbstractAttrProcessor.java:74)
    org.thymeleaf.processor.AbstractProcessor.process(AbstractProcessor.java:212)
    org.thymeleaf.dom.Node.applyNextProcessor(Node.java:805)
    org.thymeleaf.dom.Node.processNode(Node.java:767)
    org.thymeleaf.dom.NestableNode.computeNextChild(NestableNode.java:626)
    org.thymeleaf.dom.NestableNode.doAdditionalProcess(NestableNode.java:608)
    org.thymeleaf.dom.Node.processNode(Node.java:785)
    org.thymeleaf.dom.NestableNode.computeNextChild(NestableNode.java:626)
    org.thymeleaf.dom.NestableNode.doAdditionalProcess(NestableNode.java:608)
    org.thymeleaf.dom.Node.processNode(Node.java:785)
    org.thymeleaf.dom.NestableNode.computeNextChild(NestableNode.java:626)
    org.thymeleaf.dom.NestableNode.doAdditionalProcess(NestableNode.java:608)
    org.thymeleaf.dom.Node.processNode(Node.java:785)
    org.thymeleaf.dom.Document.process(Document.java:90)
    org.thymeleaf.TemplateEngine.process(TemplateEngine.java:1029)
    org.thymeleaf.TemplateEngine.process(TemplateEngine.java:978)
    org.thymeleaf.spring3.view.ThymeleafView.render(ThymeleafView.java:460)
    org.springframework.web.servlet.DispatcherServlet.render(DispatcherServlet.java:1157)
    org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:927)
    org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:827)
    org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:882)
    org.springframework.web.servlet.FrameworkServlet.doGet(FrameworkServlet.java:778)
    javax.servlet.http.HttpServlet.service(HttpServlet.java:621)
    javax.servlet.http.HttpServlet.service(HttpServlet.java:722)
```

I [asked on the forum](http://forum.thymeleaf.org/Creating-links-with-parameter-containing-td4024785.html) if there were any recommended solutions to this problem and a work around is that I can do something like this:

``` xml
<a href="#" th:href="@{/url} + '?a[0]=b'">List</a>
```

This works, but it feels like a workaround for two reasons:
1. It appears there is no way to escape the [ bracket. There should be some way to treat a[0] as a literal.
2. It requires manual escaping of the query parameters. For example if b were replaced with a ? I would need to manually URL encode the ? for it to work.

It would be nice to be able to use the standard URL mechanism even if special characters were used in the parameter names or values. I'm not particular on the syntax on how this would be supported, but one way of representing this might be something like:

``` xml
<a href="#" th:href="@{/url('a[0]'='?')}">List</a>
```

**Comment by danielfernandez** [comment]
This is probably due to Thymeleaf not accepting the "[" and "]" chars as a part of the parameter names (which it considers "tokens").

I see the usefulness of this. For example, Spring MVC uses this syntax for array/map form binding.

Will be fixed for 2.0.9.

**Comment by danielfernandez** [comment]
Solved in trunk and 2.0.9-SNAPSHOT. Will be a part of 2.0.9.

Now this syntax:

```
th:href="@{/url(a[0]='something')}"
```

Is valid.
